### PR TITLE
PP-9790 Sandbox payment provider recurring authorisation

### PR DESCRIPTION
**Pull request by sfount**
Implement user not present interface for the Sandbox payment provider.

Use the new `last4DigitsSandboxResponseGenerator` to return a gateway
response which is matched to information from the payment instrument
available to user not present authorisation.

For now we will use this generator but plan to add a subsequent
generator that will respect the existing test card numbers.

---

The charge operations task queue isn't configured to do recurring
authorisations yet. Add an integration to assert that charges with
agreement authorisation mode are seen through to authorised or declined
depending on sandbox card numbers.

With @alexbishop1

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java [hunk 1]
```diff
@@ -635,6 +635,16 @@ public ChargeEntity updateChargePostAuthorisation(String chargeExternalId,
             Optional.ofNullable(emailAddress).ifPresent(charge::setEmail);
 
             CardDetailsEntity detailsEntity = authCardDetailsToCardDetailsEntityConverter.convert(authCardDetails);
+
+            // propagate details that aren't mapped from payment instrument to auth card details onto the charge
+            // this logic should be removable when payment instruments are modelled and used for all authorisation types
+            charge.getPaymentInstrument()
+                    .ifPresent(paymentInstrument -> {
+                        if (charge.getAuthorisationMode() == AuthorisationMode.AGREEMENT) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java]: Would it make more sense to have the `if` statement wrapping the rest of the code or is there a reason it’s been done like this?

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java [hunk 2]
```diff
@@ -635,6 +635,16 @@ public ChargeEntity updateChargePostAuthorisation(String chargeExternalId,
...
+                            detailsEntity.setFirstDigitsCardNumber(paymentInstrument.getCardDetails().getFirstDigitsCardNumber());
+                            detailsEntity.setFirstDigitsCardNumber(paymentInstrument.getCardDetails().getFirstDigitsCardNumber());
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java]: One of these should be `LastDigitsCardNumber`, surely?

File: src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java [hunk 3]
```diff
@@ -50,11 +52,17 @@ public CardDetailsEntity convert(AuthCardDetails authCardDetails) {
     }
 
     private static boolean hasFullCardNumber(AuthCardDetails authCardDetails) {
-        return authCardDetails.getCardNo().length() > 6;
+        return Optional.ofNullable(authCardDetails)
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java]: I can understand there not being a card number but do we expect callers to pass a null `authCardDetails`?

File: src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java [hunk 4]
```diff
@@ -50,11 +52,17 @@ public CardDetailsEntity convert(AuthCardDetails authCardDetails) {
...
+                .map(AuthCardDetails::getCardNo)
+                .map(cardNumber -> cardNumber.length() > 6)
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java:56]: `.filter(…)` would be more appropriate than `.map(…)`.

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceIT.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        var response = testContext.getInstanceFromGuiceContainer(CardAuthoriseService.class)
+                .doAuthoriseWeb("external-charge-id", AuthCardDetailsFixture.anAuthCardDetails().build());
+        assertThat(response.getGatewayError(), is(Optional.empty()));
+        assertThat(response.getAuthoriseStatus(), is(Optional.of(BaseAuthoriseResponse.AuthoriseStatus.AUTHORISED)));
+    }
+    
+    @Test
+    public void shouldAuthoriseSandboxUserNotPresentPayment() {
+        addCharge("success-charge-external-id", "4242");
+        addCharge("decline-charge-external-id", "0002");
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceIT.java]: I’m lazy so we can add constants for `"4242"` and `"0002"` describing what outcomes they will trigger?

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceIT.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        var successCharge = testContext.getInstanceFromGuiceContainer(ChargeService.class).findChargeByExternalId("success-charge-external-id");
+        var declineCharge = testContext.getInstanceFromGuiceContainer(ChargeService.class).findChargeByExternalId("decline-charge-external-id");
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceIT.java]: Might be better to split this into two tests?

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
Looks good, just a couple of things.

**Review by alexbishop1 — APPROVED** [review]

File: src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java (hunk 3 above)
Comment (sfount) [reply src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java]: I've opted to use `map` for the moment as I think it most helpfully describes taking a optional string value and returning a boolean. `filter` `ifPresent()` would be slick though, I hadn't thought of that

File: src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java (hunk 4 above)
Comment (sfount) [reply src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java:56]: I've opted to use `map` for the moment as I think it most helpfully describes taking one optional string value and returning a boolean. `filter` `ifPresent()` would be slick though, I hadn't thought of that

**Pull request merged**